### PR TITLE
fix(PPDSC-2605): safari svg filters fix

### PR DESCRIPTION
**Pull request by evgenitsn**
PPDSC-2605

**What**

This fixes the issue with Safari where some of the SVG are not rendered correctly.
The issue is that Safari does not support additional filters like shadows, blurs and more if the svg is external i.e.imported using `<use href={...} />`.

The workaround for this is to create a new empty `<svg>` and add the `<defs>` section of the svgs (the part where the filters are located) and add this svg anywhere in the DOM.

What I did is to check if the browser is Safari, and if it is to read the svg as raw string, sanitize it, get substring of that `<defs> .. </defs>` and append it as empty, hidden svg next to the original svg. 

If the browser is not Safari, this logic is skipped and nothing changes. 
Looks like this approach fixes the Safari issues with the illustrations.

Example of the DOM for Safari vs Chrome

Safari:
image

Chrome:
image

**I have done:**
- [ ] Written unit tests against changes
- [ ] Written functional tests against the component and/or NewsKit site
- [ ] Updated relevant documentation

**I have tested manually:**
- [x] The feature's functionality is working as expected on Chrome, Firefox, Safari and Edge
- [ ] The screen reader reads and flows through the elements as expected.
- [x] There are no new errors in the browser console coming from this PR.
- [ ] When visual test is not added, it renders correctly on different browsers and mobile viewports (Safari, Firefox, small mobile viewport, tablet)
- [ ] The Playground feature is working as expected

**Before:**

image

**After:**
image

**Who should review this PR:**

**How to test:**

### CODE REVIEW

**Comment by newskit-engineering** [comment]
You can preview these changes on:
  Documentation Site: http://ncu-newskit-docs.s3-website-eu-west-1.amazonaws.com/ppdsc-2605-svg-shadow-elements-missing-in-safari
  Storybook: http://ncu-newskit-docs.s3-website-eu-west-1.amazonaws.com/ppdsc-2605-svg-shadow-elements-missing-in-safari/storybook

File: package.json [hunk 1]
```diff
@@ -188,6 +188,7 @@
     "package-change-checker": "^1.2.4",
     "prettier": "2.2.1",
     "pretty-proptypes": "^0.6.0",
+    "raw-loader": "^4.0.2",
```
Comment (mutebg) [inline package.json:191]: Do we need to add this to ignore dependencies list?

File: site/components/illustrations/illustration-loader.tsx [hunk 2]
```diff
@@ -1,5 +1,34 @@
...
-import {pathToID} from './utils';
+import dompurify from 'isomorphic-dompurify';
+import {isSafari, pathToID} from './utils';
+
+/**
+ * Only for Safari. It reads the svg file as raw string and returns the problematic <defs> .. </defs> part as additional svg which solves the problem. If the browser is not Safari then all of this is skipped.
+ */
+const loadSVGFiltersForSafari = (path: string) => {
+  if (!isSafari) return null;
+  const {sanitize} = dompurify;
```
Comment (mutebg) [inline site/components/illustrations/illustration-loader.tsx:10]: wondering if we can load the dompurify with dynamic import so that our bundle size does not increase

File: package.json (hunk 1 above)
Comment (evgenitsn) [reply package.json:191]: This is a `devDependency` I believe it will be omitted automatically. The `dependenciesNotToBeIncluded` list does contains only `dependencies` and not `devDependency`. Let me know if I am missing something.

File: site/components/illustrations/illustration-loader.tsx (hunk 2 above)
Comment (evgenitsn) [reply site/components/illustrations/illustration-loader.tsx:10]: In this case we can, but we are already using the dompurify in another Component: `site/components/tools/svg-previewer.tsx`. So I think the bundle size is already including that. 

I tried to import it dynamically there as well but I wasn't able to figure out how to dynamically import an npm package in a react component body. If you have any ideas, happy to discuss.

Comment (mutebg) [reply site/components/illustrations/illustration-loader.tsx:10]: there is a difference I believe, there is a bundle per page, so the previewer will include it only there, and since we start using it for SVG's and they are on all pages it will start to be included everywhere.
I think you might need to do that inside useEffect and maybe create a custom hook in order to achieve it. I am not sure if its with the time these extra 10kb but have a try

File: package.json (hunk 1 above)
Comment (mutebg) [reply package.json:191]: You are right, I did not see it

File: site/components/illustrations/illustration-loader.tsx (hunk 2 above)
Comment (evgenitsn) [reply site/components/illustrations/illustration-loader.tsx]: Updated the code, the sanitizer should be load dynamically only for Safari now.

File: site/components/illustrations/illustration-loader.tsx [hunk 3]
```diff
@@ -1,18 +1,78 @@
...
+const getSafariSVGFilters = (
+  sanitize: (source: string) => string,
+  path: string,
+) => {
+  const rawSVG = sanitize(
+    // eslint-disable-next-line import/no-dynamic-require, global-require
+    require(`../../public/static/illustrations/${path}.svg`).default,
+  );
+
+  const defsStart = '<defs>';
```
Comment (mstuartf) [inline site/components/illustrations/illustration-loader.tsx:16]: Could there be any attributes in this opening tag that might mean the match doesn't work?

File: site/components/illustrations/illustration-loader.tsx [hunk 4]
```diff
@@ -1,18 +1,78 @@
...
   const id = pathToID(path);
 
-  const Component = () => (
+  const renderInlineSVG = () => (
     <svg viewBox="0 0 1490 838" width="100%" {...props}>
       <use href={`static/illustrations/${path}.svg#${id}`} />
     </svg>
   );
 
+  const Component = () => {
```
Comment (mstuartf) [inline site/components/illustrations/illustration-loader.tsx:44]: Can we add tests for this logic?

File: site/components/illustrations/illustration-loader.tsx (hunk 3 above)
Comment (LukeFinch) [reply site/components/illustrations/illustration-loader.tsx:16]: <defs> can have "lang" and "id" they're also technically styleable.
Might be worth doing a regex to match defs just in case the export from figma adds anything in there.

`<defs[^(\/>)]*>` this should catch any opening def tag

**Comment by mstuartf** [comment]
Screenshot 2022-11-25 at 10 39 18
Screenshot 2022-11-25 at 10 39 22
@evgenitsn fyi it looks like this is being picked up by the Percy tests, so you will have some to resolve.

**Comment by evgenitsn** [comment]
> Screenshot 2022-11-25 at 10 39 18
> 
> Screenshot 2022-11-25 at 10 39 22
> 
> @evgenitsn fyi it looks like this is being picked up by the Percy tests, so you will have some to resolve.

Thanks Mike, yes I noticed this in a few more cases. I will push an update. Marking it as WIP now as the implementation would change a bit.

**Comment by evgenitsn** [comment]
I've updated the implementation, it turned out that appending only the <defs> </defs> section is not reliable solution because not all filters are contained inside it. 

So currently for Safari we are just inlining the SVG into the code (same as before the svg/tsx change) and for the rest of the browsers we are using external SVGs inside the <use/> i.e. how it is supposed to work.

File: site/components/illustrations/illustration-loader.tsx (hunk 4 above)
Comment (evgenitsn) [reply site/components/illustrations/illustration-loader.tsx:34]: I tried to add testing to this, but I am facing issues with the jest SVG tranform and also I get weird errors which I guess are because of the dynamic import because it's an async operation so I am a bit stuck at the moment.

If anyone keen to assist we can try to solve these. I have a feeling that this task already took a lot of time so idk :/

**Comment by evgenitsn** [comment]
Also there is one failing e2e test which on a first looks shouldn't be involved with my changes. I am having issues reproducing it locally.

File: site/components/illustrations/illustration-loader.tsx (hunk 3 above)
Comment (evgenitsn) [reply site/components/illustrations/illustration-loader.tsx]: Thanks @LukeFinch I left a comment below that this `<defs..>` section is not enough for all illustrations so I've updated the code to inject inline the whole svg only for Safari. So this comment is not relevant anymore.

File: site/components/illustrations/illustration-loader.tsx (hunk 4 above)
Comment (mstuartf) [reply site/components/illustrations/illustration-loader.tsx:34]: OK, maybe we can add in the future

**Comment by evgenitsn** [comment]
@mstuartf @mutebg @LukeFinch Thank for looking into this. I think it's in a good shape now, ready for reviews.

File: site/components/illustrations/utils.ts [hunk 5]
```diff
@@ -33,3 +33,11 @@ export const pathToID = (path: string) => {
...
   return '';
 };
+
+const safariUACheck = /^((?!chrome|android).)*safari/i;
+const isWindowLoaded = typeof window !== 'undefined';
+
+export const isSafari =
+  isWindowLoaded && safariUACheck.test(window.navigator.userAgent);
+export const isNotSafari =
+  isWindowLoaded && !safariUACheck.test(window.navigator.userAgent);
```
Comment (LukeFinch) [inline site/components/illustrations/utils.ts:43]: What's the reason for having both? does `!isSafari` not satisfy the need?

Comment (LukeFinch) [reply site/components/illustrations/utils.ts:43]: Fair enough, I figured it would have to do with the window being loaded. 👍

**Review by LukeFinch — DISMISSED** [review]

File: site/components/illustrations/illustration-loader.tsx [hunk 6]
```diff
@@ -1,17 +1,69 @@
...
+        viewBox="0 0 1490 838"
+        width="100%"
+        opacity={isNotSafari ? 1 : 0}
+        {...props}
+      >
+        <use href={`static/illustrations/${path}.svg#${id}`} />
+      </svg>
+    );
+
+    return safariSVG === null ? nonSafariSVG : renderSafariSVG;
```
Comment (mstuartf) [inline site/components/illustrations/illustration-loader.tsx:65]: Can this be simplified?

I think there is duplicate logic:
* `safariSVG` will be null unless `isSafari` is true due to the if condition in the `useEffect`
* But you are also checking `isSafari` again when defining `renderSafariSVG`

Could you just remove `renderSafariSVG` and make the final return statement:
`return safariSVG || nonSafariSVG;`

Or do you want `nonSafariSVG` to render first, then to be replaced with `safariSVG` when the async logic finishes?

File: site/components/illustrations/illustration-loader.tsx [hunk 7]
```diff
@@ -1,17 +1,69 @@
...
+        // Dynamically importing 'isomorphic-dompurify' only for Safari to avoid bloating the bundle size for other browsers.
+        const importSanitizer = async () => {
+          const dompurify = await (await import('isomorphic-dompurify'))
+            .default;
+          const svg = getSafariSVG(dompurify.sanitize, path, props);
+          setSafariSVG(svg);
+        };
+        importSanitizer();
+      }
+    }, [safariSVG]);
```
Comment (mstuartf) [inline site/components/illustrations/illustration-loader.tsx:48]: Could this just run on first render with `[]` as dependencies, and no need to check `safariVG === null`?

File: site/components/illustrations/illustration-loader.tsx (hunk 6 above)
Comment (evgenitsn) [reply site/components/illustrations/illustration-loader.tsx]: Just tested this one, it works fine and it looks much cleaner, thanks. Some of these checks were leftover from the previous approach using `<defs>` substring because we had to have more control over the order of rendering but with the current approach we can simplify. 👍

File: site/components/illustrations/illustration-loader.tsx (hunk 7 above)
Comment (evgenitsn) [reply site/components/illustrations/illustration-loader.tsx]: Done

**Review by mstuartf — DISMISSED** [review]

**Review by LukeFinch — APPROVED** [review]

**Review by LukeFinch — DISMISSED** [review]

File: site/components/tools/svg-previewer.tsx [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import React, {useEffect, useState} from 'react';
 import {getColorCssFromTheme, getSSRId, P} from 'newskit';
-import dompurify from 'dompurify';
+import dompurify from 'isomorphic-dompurify';
```
Comment (Vanals) [inline site/components/tools/svg-previewer.tsx:3]: Hey! why is this needed 🤔 ?

Comment (evgenitsn) [reply site/components/tools/svg-previewer.tsx:3]: If you are asking if the library change from `dompurify` to `isomorphic-dompurify` is needed. I added that because I need `isomophic-dompurify` into another component and decided to swap it in both places to not keep two libs for almost the same thing.

If you are asking if we need dompurify lib for this `svg-previewer.tsx` file, I am not sure, I haven't spend much time looking into this file and what it is doing, just replaced the lib. So can't really tell.

**Review by LukeFinch — APPROVED** [review]

**Review by Vanals — APPROVED** [review]
